src/server/raft_client.rs

```diff
@@ -11,192 +11,336 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::VecDeque;
+use std::collections::hash_map::Entry;
 use std::ffi::CString;
+use std::fmt::{self, Display, Formatter};
+use std::mem;
 use std::sync::Arc;
-use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering, ATOMIC_USIZE_INIT};
+use std::sync::atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT};
 
+use futures::future::{self, Shared, SharedError};
 use futures::sync::mpsc::{self, UnboundedSender};
 use futures::sync::oneshot::{self, Sender};
-use futures::{stream, Future, Sink, Stream};
+use futures::{stream, Async, Future, Poll, Sink, Stream};
+use futures_cpupool::{Builder, CpuPool};
 use grpc::{ChannelBuilder, Environment, WriteFlags};
 use kvproto::raft_serverpb::RaftMessage;
 use kvproto::tikvpb_grpc::TikvClient;
 
 use super::metrics::*;
+use super::resolve::{ResolveResult, StoreAddrResolver};
 use super::{Config, Error, Result};
 use util::collections::HashMap;
 use util::security::SecurityManager;
 
 const MAX_GRPC_RECV_MSG_LEN: usize = 10 * 1024 * 1024;
 const MAX_GRPC_SEND_MSG_LEN: usize = 10 * 1024 * 1024;
-const PRESERVED_MSG_BUFFER_COUNT: usize = 1024;
+const PRESERVED_MSG_BUFFER_COUNT: usize = 1024 - 1;
```

Comment:
why  -1 here?

Reply:
`VecDeque` adds another slot for giving capacity.